k8s/pkg/prefixcollector/service.go

```diff
@@ -1,112 +1,97 @@
 package prefixcollector
 
 import (
-	"context"
+	"strings"
 	"sync"
 
-	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection"
-
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 
-	"github.com/networkservicemesh/networkservicemesh/controlplane/api/plugins"
+	"github.com/networkservicemesh/networkservicemesh/sdk/common"
 	"github.com/networkservicemesh/networkservicemesh/sdk/prefix_pool"
 )
 
+const (
+	configMapName = "nsm-config"
+)
+
 type prefixService struct {
 	sync.RWMutex
 	excludedPrefixes prefix_pool.PrefixPool
 }
 
-// NewPrefixService creates an instance of ConnectionPluginServer
-func NewPrefixService(config *rest.Config) (plugins.ConnectionPluginServer, error) {
+// NewExcludePrefixServer creates an instance of prefixService
+func NewExcludePrefixServer(config *rest.Config) error {
 	clientset, err := kubernetes.NewForConfig(config)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	emptyPrefixPool, err := prefix_pool.NewPrefixPool()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	rv := &prefixService{
 		excludedPrefixes: emptyPrefixPool,
 	}
 
 	if err := rv.monitorExcludedPrefixes(clientset); err != nil {
-		return nil, err
+		return err
 	}
 
-	return rv, nil
+	return nil
 }
 
-func (c *prefixService) getExcludedPrefixes() prefix_pool.PrefixPool {
-	c.RLock()
-	defer c.RUnlock()
+func (ps *prefixService) getExcludedPrefixes() prefix_pool.PrefixPool {
+	ps.RLock()
+	defer ps.RUnlock()
 
-	return c.excludedPrefixes
+	return ps.excludedPrefixes
 }
 
-func (c *prefixService) setExcludedPrefixes(prefixes prefix_pool.PrefixPool) {
-	c.Lock()
-	defer c.Unlock()
+func (ps *prefixService) setExcludedPrefixes(prefixes prefix_pool.PrefixPool) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	ps.excludedPrefixes = prefixes
+}
 
-	c.excludedPrefixes = prefixes
+func buildPrefixesYaml(prefixes []string) string {
+	marker := "prefixes:"
+	if len(prefixes) == 0 {
+		return marker + " []"
+	}
+	return strings.Join(append([]string{marker}, prefixes...), "\n- ")
 }
 
-func (c *prefixService) monitorExcludedPrefixes(clientset *kubernetes.Clientset) error {
+func (ps *prefixService) monitorExcludedPrefixes(clientset *kubernetes.Clientset) error {
 	poolCh, err := getExcludedPrefixesChan(clientset)
 	if err != nil {
 		return err
 	}
 
 	go func() {
+		configMaps := clientset.CoreV1().ConfigMaps(common.GetNamespace())
 		for pool := range poolCh {
-			c.setExcludedPrefixes(pool)
+			prefixes := pool.GetPrefixes()
+			logrus.Infof("Excluded prefixes changed: %v", prefixes)
+			cm, err := configMaps.Get(configMapName, metav1.GetOptions{})
+			if err != nil {
+				logrus.Errorf("Failed to get ConfigMap '%s/%s': %v", common.GetNamespace(), configMapName, err)
+				continue
```

Comment:
You have some latest prefixes in `prefixes` and there is probability that they won't change for a long period of time. And after you failed to get ConfigMap it doesn't look like good idea to call `continue` - you might block for unpredictable period of time

Reply:
- [ ] Make saving of latest prefixes persistent